### PR TITLE
[Block-STM] Incorporate delta materialization with rolling commit

### PR DESCRIPTION
**Pull request by gelash**
Removing two bottlenecks for opening up the aggregators:
(1) parallel processing of different aggregators / materialization of deltas: this way, having many aggregators in the system shouldn't hurt performance due to some sequential bottleneck. 
(2) optimize aggregator reads so they aren't prohibitively expensive - previously they could traverse all deltas, now the traversal will stop at the rolling commit head. So read is still undesired and expensive, but not prohibitively expensive.

How is this accomplished:
- We get rid of OutputDeltaResolver, instead every time an index is committed by the coordinator thread, commit_hook is executed that materializes the deltas produced by the last incarnation of this transaction and records them in TxnLastInputOutput as well.
- The coordinator assigns running commit hooks to different worker threads, and we make sure commit_hook is executed for all committed transactions as the threads can't break until they receive all txn indices for commit hook.
- In the multi-version data-structure we can get rid of all the crust that OutputDeltaResolver needed. Instead, we add shortcut for base value of an aggregator (so we don't try to resolve from storage many times), and shortcut for a committed value to inner entries (stored once the corresponding txn is committed).

@georgemitenkov @sasha @zekun000 let's remember to benchmark these claims - shouldn't affect landing the PR since it certainly improves things - i.e. this is necessary, and hopefully if we aren't missing something, also sufficient to provide (1) and (2) above.

### CODE REVIEW

File: aptos-move/mvhashmap/src/versioned_data.rs [hunk 1]
```diff
@@ -294,4 +269,28 @@ impl<K: Hash + Clone + Eq, V: TransactionWrite> VersionedData<K, V> {
...
+        match v.read(txn_idx + 1) {
+            Ok(MVDataOutput::Resolved(value)) => {
+                v.versioned_map
+                    .get_mut(&txn_idx)
+                    .expect("Entry by the txn must exist to commit delta")
+                    .record_delta_shortcut(value);
+
+                Ok(value)
+            },
+            Err(MVDataError::Unresolved(_)) => bail!("No base value"),
```
Comment (georgemitenkov) [inline aptos-move/mvhashmap/src/versioned_data.rs:292]: Why the case when `read()` returns `Unresolved(..)` cannot happen here? If we have the first txn which we commit then the delta will not have any shortcuts, and so returns an unresolved call to go to storage?

File: aptos-move/mvhashmap/src/versioned_data.rs [hunk 2]
```diff
@@ -294,4 +269,28 @@ impl<K: Hash + Clone + Eq, V: TransactionWrite> VersionedData<K, V> {
...
         }));
     }
+
+    // When a transaction is committed, this method can be called for its delta outputs to add
+    // a 'shortcut' to the corresponding materialized aggregator value, so any subsequent reads
+    // do not have to traverse below the index. It must be guaranteed by the caller that the
+    // data recorded below this index will not change after the call, and that the corresponding
+    // transaction has indeed produced a delta recorded at the given key.
+    // If the result is err, it means the base value hadn't been set.
+    pub(crate) fn commit_delta(&self, key: &K, txn_idx: TxnIndex) -> anyhow::Result<u128> {
```
Comment (georgemitenkov) [inline aptos-move/mvhashmap/src/versioned_data.rs:279]: Maybe option instead of result here?

File: aptos-move/block-executor/src/executor.rs [hunk 3]
```diff
@@ -199,6 +213,37 @@ where
...
+        txn_idx: TxnIndex,
+        versioned_cache: &MVHashMap<T::Key, T::Value, ExecutableTestType>,
+        last_input_output: &TxnLastInputOutput<T::Key, E::Output, E::Error>,
+        base_view: &S,
+    ) {
+        let delta_keys = last_input_output.delta_keys(txn_idx);
+        let mut delta_writes = Vec::with_capacity(delta_keys.len());
+        for k in delta_keys {
+            let mut committed_delta = versioned_cache.commit_delta(&k, txn_idx);
+            if committed_delta.is_err() {
```
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:227]: @georgemitenkov it's handled here on the caller side?

File: aptos-move/block-executor/src/executor.rs [hunk 4]
```diff
@@ -207,27 +252,42 @@ where
...
+                        post_commit_txs[idx]
+                            .send(txn_idx)
+                            .expect("Worker must be available");
+                        // Iterate round robin over workers to do commit_hook.
+                        idx = (idx + 1) % post_commit_txs.len();
+
+                        if txn_idx as usize + 1 == block.len() {
+                            // Committed the last transaction / everything.
+                            scheduler_task = SchedulerTask::Done;
+                            break;
```
Comment (zekun000) [inline aptos-move/block-executor/src/executor.rs:280]: this might be incorrect if there's only one worker, (it breaks without draining the queue), we don't spawn one worker for execution today but seems like a bomb

File: aptos-move/block-executor/src/executor.rs [hunk 5]
```diff
@@ -279,21 +351,44 @@ where
 
         let num_txns = signature_verified_block.len() as u32;
         let last_input_output = TxnLastInputOutput::new(num_txns);
-        let committing = AtomicBool::new(true);
         let scheduler = Scheduler::new(num_txns);
 
+        let roles: Vec<CachePadded<Mutex<Option<CommitRole>>>> = (0..self.concurrency_level)
```
Comment (zekun000) [inline aptos-move/block-executor/src/executor.rs:356]: why this is a mutex of option? isn't Vec<CommitRole> enough? 
```
        let mut roles: Vec<CommitRole> = vec![];
        let mut senders = Vec::with_capacity(self.concurrency_level - 1);
        for _ in 0..(self.concurrency_level - 1) {
            let (tx, rx) = mpsc::channel();
            roles.push(CommitRole::Worker(rx));
            senders.push(tx);
        }
        roles.push(CommitRole::Coordinator(senders, 0));
```

File: aptos-move/block-executor/src/scheduler.rs [hunk 6]
```diff
@@ -367,25 +367,28 @@ impl Scheduler {
...
-        self.suspend(txn_idx, dep_condvar.clone());
+            // Safe to add dependency here (still holding the lock) - finish_execution of txn
+            // dep_txn_idx is guaranteed to acquire the same lock later and clear the dependency.
+            stored_deps.push(txn_idx);
 
-        // Safe to add dependency here (still holding the lock) - finish_execution of txn
-        // dep_txn_idx is guaranteed to acquire the same lock later and clear the dependency.
-        stored_deps.push(txn_idx);
+            // Stored deps gets unlocked here.
+        }
```
Comment (zekun000) [inline aptos-move/block-executor/src/scheduler.rs:391]: why this change is needed?

File: aptos-move/block-executor/src/txn_last_input_output.rs [hunk 7]
```diff
@@ -231,15 +222,58 @@ impl<K: ModulePath, T: TransactionOutput, E: Send + Clone> TxnLastInputOutput<K,
         }
     }
 
+    pub(crate) fn delta_keys(&self, txn_idx: TxnIndex) -> KeySet<T> {
```
Comment (zekun000) [inline aptos-move/block-executor/src/txn_last_input_output.rs:225]: nit: this doesn't feel it requires a new HashSet, seems an Iterator is enough

File: aptos-move/block-executor/src/executor.rs (hunk 4 above)
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:280]: We have an assert for it? I can add a comment to the assert that it wouldn't actually work? (and it would be not too hard to fix if it was ever needed somehow - just calling hook itself.).

**Comment by davidiw** [comment]
> let's remember to benchmark these claims - shouldn't affect landing the PR since it certainly improves things

Curious, why we don't want to benchmark non-trivial changes before landing? I'd love to know what the impact of aggregators are now before and after this. Would the executor benchmark with gas using the aggregator give us some signal?

**Comment by gelash** [comment]
https://github.com/aptos-labs/aptos-core/pull/7773/files to benchmark w. aggregator total supply before we can land.

File: aptos-move/block-executor/src/scheduler.rs (hunk 6 above)
Comment (gelash) [reply aptos-move/block-executor/src/scheduler.rs:391]: was just to unlock asap, but it's not too important, can remove.

File: aptos-move/block-executor/src/executor.rs [hunk 8]
```diff
@@ -207,27 +252,42 @@ where
...
+                        if txn_idx as usize + 1 == block.len() {
+                            // Committed the last transaction / everything.
+                            scheduler_task = SchedulerTask::Done;
+                            break;
+                        }
                     }
-                }
+                },
+                CommitRole::Worker(rx) => {
+                    while let Ok(txn_idx) = rx.try_recv() {
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:285]: Hope is that since threads finish their regular tasks at different times, even if the coordinator observed many commits at the same time and dispatched them, their commit_hooks wouldn't necessarily be executed concurrently. Moreover, once we have different aggregators, that should spread out contention as well. So hopefully happens few enough times, and also the contention case itself is a relatively efficient lock processing.

Since the total supply aggregator is a good example here (all txns touch it), once we have https://github.com/aptos-labs/aptos-core/pull/7773/files, we should make sure we don't see a regression for even larger number of cores (@danielxiangzl )

File: aptos-move/block-executor/src/executor.rs [hunk 9]
```diff
@@ -199,6 +213,37 @@ where
...
+        &self,
+        txn_idx: TxnIndex,
+        versioned_cache: &MVHashMap<T::Key, T::Value, ExecutableTestType>,
+        last_input_output: &TxnLastInputOutput<T::Key, E::Output, E::Error>,
+        base_view: &S,
+    ) {
+        let delta_keys = last_input_output.delta_keys(txn_idx);
+        let mut delta_writes = Vec::with_capacity(delta_keys.len());
+        for k in delta_keys {
+            let mut committed_delta = versioned_cache.commit_delta(&k, txn_idx);
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:226]: done

File: aptos-move/block-executor/src/txn_last_input_output.rs [hunk 10]
```diff
@@ -231,15 +222,58 @@ impl<K: ModulePath, T: TransactionOutput, E: Send + Clone> TxnLastInputOutput<K,
...
+        match &self.outputs[txn_idx as usize].load_full() {
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/txn_last_input_output.rs:226]: load should be fine, done

**Comment by danielxiangzl** [comment]
Just tested the execution-only performance of the PR. It looks good to me. No degradation with 16 threads and only a slight degradation (~1k) with 32 threads.

File: aptos-move/block-executor/src/executor.rs [hunk 11]
```diff
@@ -199,6 +210,40 @@ where
...
+        txn_idx: TxnIndex,
+        versioned_cache: &MVHashMap<T::Key, T::Value, ExecutableTestType>,
+        last_input_output: &TxnLastInputOutput<T::Key, E::Output, E::Error>,
+        base_view: &S,
+    ) {
+        let (num_deltas, delta_keys) = last_input_output.delta_keys(txn_idx);
+        let mut delta_writes = Vec::with_capacity(num_deltas);
+        for k in delta_keys {
+            let committed_delta = versioned_cache
+                .materialize_delta(&k, txn_idx)
```
Comment (zekun000) [inline aptos-move/block-executor/src/executor.rs:224]: this is a bit ugly in the sense that the first time it's guaranteed to fail without storage value and needs to read and re-materialize again, maybe we should just check if aggregator_base is set and set it before calling materialize?

File: aptos-move/block-executor/src/txn_last_input_output.rs [hunk 12]
```diff
@@ -5,24 +5,40 @@ use crate::{
...
 };
 
 type TxnInput<K> = Vec<ReadDescriptor<K>>;
-type TxnOutput<T, E> = ExecutionStatus<T, Error<E>>;
+// When a transaction is committed, the output delta writes must be populated by
+// the WriteOps corresponding to the deltas in the corresponding outputs.
+struct TxnOutput<T: TransactionOutput, E> {
+    output_status: ExecutionStatus<T, Error<E>>,
+    materialized_deltas:
+        OnceCell<Vec<(<<T as TransactionOutput>::Txn as Transaction>::Key, WriteOp)>>,
```
Comment (zekun000) [inline aptos-move/block-executor/src/txn_last_input_output.rs:29]: ditto on the T::Txn::Key

**Review by zekun000 — APPROVED** [review]

**Review by danielxiangzl — APPROVED** [review]

File: aptos-move/block-executor/src/txn_last_input_output.rs [hunk 13]
```diff
@@ -231,17 +233,72 @@ impl<K: ModulePath, T: TransactionOutput, E: Send + Clone> TxnLastInputOutput<K,
...
+                Box::new(empty::<<<T as TransactionOutput>::Txn as Transaction>::Key>()),
+            ),
+            |txn_output| match &txn_output.output_status {
+                ExecutionStatus::Success(t) | ExecutionStatus::SkipRest(t) => {
+                    let deltas = t.get_deltas();
+                    return (deltas.len(), Box::new(deltas.into_iter().map(|(k, _)| k)));
+                },
+                ExecutionStatus::Abort(_) => (
+                    0,
+                    Box::new(empty::<<<T as TransactionOutput>::Txn as Transaction>::Key>()),
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/txn_last_input_output.rs:255]: I get errors like:
 ambiguous associated type
   --> aptos-move/block-executor/src/txn_last_input_output.rs:235:42
    |
235 |     ) -> (usize, Box<dyn Iterator<Item = T::Txn::Key>>) {
    |                                          ^^^^^^^^^^^ help: use fully-qualified syntax: `<<T as task::TransactionOutput>::Txn as Trait>::Key`

File: aptos-move/block-executor/src/executor.rs (hunk 11 above)
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:224]: Issue is overall it will be way less efficient because the base needs to be set once and materializing doesn't necessarily need to check it, most of the time it will just succeed based on some earlier materialized write. It's just when it fails without resolving we know base must have been missing and we can set it.. In other words, if we always checked, we would do ~2 times more locks overall than now - may fail in the beginning, but later everything works without even getting to the aggregator base.

**Comment by gelash** [comment]
Responded to comments.

Also added to trait so we can incorporate delta outputs already during commit_hook, simplifying interfaces and the flow (no need to use executor threadpool for postprocessing).

If you get a chance, please have another quick look before merging? @zekun000 @danielxiangzl

File: aptos-move/aptos-vm/src/block_executor/mod.rs [hunk 14]
```diff
@@ -39,47 +41,80 @@ impl BlockExecutorTransaction for PreprocessedTransaction {
...
+    output_with_deltas: OnceCell<TransactionOutput>,
+}
 
 impl AptosTransactionOutput {
-    pub fn new(output: TransactionOutputExt) -> Self {
-        Self(output)
+    pub(crate) fn new(output: TransactionOutputExt) -> Self {
+        Self {
+            output_ext: Mutex::new(Some(output)),
+            output_with_deltas: OnceCell::new(),
```
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/block_executor/mod.rs:54]: Can we maybe call it `materialized_output` or `committed_output`? To me `output_with_deltas` sounds like there is and output and it has deltas separately

File: aptos-move/block-executor/src/executor.rs [hunk 15]
```diff
@@ -199,6 +210,50 @@ where
...
+                .unwrap_or_else(|| {
+                    let storage_value = base_view
+                        .get_state_value_bytes(&k)
+                        .expect("No base value for committed delta in storage")
+                        .map(|bytes| deserialize(&bytes))
+                        .expect("Cannot deserialize base value for committed delta");
+
+                    versioned_cache.set_aggregator_base_value(&k, storage_value);
+                    versioned_cache
+                        .materialize_delta(&k, txn_idx)
```
Comment (georgemitenkov) [inline aptos-move/block-executor/src/executor.rs:244]: Why not make materialization return Result and wrap Ok(u128), Err(DeltaOp) ? Then we do not need to re-materialize (which does a read and a delta traversal), and can apply delta instead?

File: aptos-move/block-executor/src/executor.rs (hunk 11 above)
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:234]: done, according to George's suggestion, no need to materialize twice.
